django/db/models/query_utils.py

```diff
@@ -99,6 +99,45 @@ def resolve_expression(
         query.promote_joins(joins)
         return clause
 
+    def replace_expressions(self, replacements):
+        if not replacements:
+            return self
+        clone = self.create(connector=self.connector, negated=self.negated)
+        for child in self.children:
+            child_replacement = child
+            if isinstance(child, tuple):
+                lhs, rhs = child
+                if LOOKUP_SEP in lhs:
+                    path, lookup = lhs.rsplit(LOOKUP_SEP, 1)
+                else:
+                    path = lhs
+                    lookup = None
+                field = models.F(path)
+                if (
+                    field_replacement := field.replace_expressions(replacements)
+                ) is not field or lookup is None:
+                    # Handle the implicit __exact case by falling back to an extra
+                    # transform when get_lookup returns no match for the last
+                    # component of the path.
+                    if lookup is None:
+                        lookup = "exact"
+                    if (lookup_class := field_replacement.get_lookup(lookup)) is None:
+                        if (
+                            transform_class := field_replacement.get_transform(lookup)
+                        ) is not None:
+                            field_replacement = transform_class(field_replacement)
+                            lookup = "exact"
+                            lookup_class = field_replacement.get_lookup(lookup)
+                    if rhs is None and lookup == "exact":
+                        lookup_class = field_replacement.get_lookup("isnull")
+                        rhs = True
+                    if lookup_class is not None:
+                        child_replacement = lookup_class(field_replacement, rhs)
```

Comment:
There might be ways to re-organize the logic to be less convoluted here. But I think the basic properties this needs is

1. We should delegate as much as possible to `F.replace_expressions`
2. We should handle implicit `__exact` lookups
3. We should avoid replacement if there are invalid lookups or transforms (`Field.replace_expressions` will return `self` in these cases).
4. We should support the no-transform but lookup cases as well (with or without an explicit `__exact`; `Q(timestamp=value)` and `Q(timestamp__gte=value)`)